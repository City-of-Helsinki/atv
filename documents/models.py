from django.conf import settings
from django.contrib.postgres.indexes import GinIndex
from django.core.serializers.json import DjangoJSONEncoder
from django.db import models
from django.urls import reverse
from django.utils.translation import gettext_lazy as _
from encrypted_fields import fields

from atv.exceptions import MaximumFileSizeExceededException
from documents.utils import get_attachment_file_path
from documents.validators import BusinessIDValidator
from services.models import Service
from utils.models import TimestampedModel, UUIDModel


class Attachment(TimestampedModel):
    document = models.ForeignKey(
        "Document",
        on_delete=models.CASCADE,
        verbose_name=_("attachment"),
        related_name="attachments",
        help_text=_("Document to which the file is attached."),
    )
    media_type = models.CharField(
        max_length=255,
        default="",
        verbose_name=_("media type"),
        help_text=_("The media type of the attachment."),
    )
    size = models.PositiveIntegerField(
        verbose_name=_("size"),
        help_text=_("Size of the attachment in bytes."),
    )
    filename = models.CharField(
        max_length=255,
        verbose_name=_("filename"),
        help_text=_("The original filename of the attachment."),
    )
    # TODO: The file is not being encrypted yet
    file = models.FileField(
        upload_to=get_attachment_file_path,
        verbose_name=_("file"),
        help_text=_("Encrypted file."),
    )

    class Meta:
        verbose_name = _("attachment")
        verbose_name_plural = _("attachments")
        default_related_name = "attachments"

    def __str__(self):
        return f"Attachment {self.pk}"

    @property
    def uri(self):
        return reverse("documents-attachments-detail", args=[self.document.id, self.id])

    def clean(self):
        if self.file.size > settings.MAX_FILE_SIZE:
            raise MaximumFileSizeExceededException()

    def save(self, *args, **kwargs):
        self.size = self.file.size
        self.filename = self.file.name

        self.full_clean()

        super().save(*args, **kwargs)


class Document(UUIDModel, TimestampedModel):
    service = models.ForeignKey(
        Service,
        on_delete=models.PROTECT,
        verbose_name=_("service"),
        help_text=_("The service which owns this document."),
    )
    user = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        blank=True,
        null=True,
        on_delete=models.PROTECT,
        verbose_name=_("user"),
        help_text=_("The creator/owner of this document."),
    )
    business_id = models.CharField(
        max_length=9,
        blank=True,
        validators=[BusinessIDValidator()],
        verbose_name=_("business ID"),
        help_text=_("The business ID of the organization which owns this document."),
    )
    transaction_id = models.CharField(
        max_length=255,
        blank=True,
        verbose_name=_("transaction ID"),
        help_text=_(
            "Transaction identifier given by the owning service. Could be e.g. a UUID."
        ),
    )
    tos_function_id = models.CharField(
        max_length=32,
        verbose_name=_("TOS function ID"),
        help_text=_(
            "UUID without dashes. Should correspond with a Function instance "
            "(e.g. the id from https://api.hel.fi/helerm/v1/function/eb30af1d9d654ebc98287ca25f231bf6/) "
            "which is applied to the stored document when considering storage time."
        ),
    )
    tos_record_id = models.CharField(
        max_length=32,
        verbose_name=_("TOS record ID"),
        help_text=_(
            "UUID without dashes. Should correspond to a record ID "
            "(e.g. records[].id from https://api.hel.fi/helerm/v1/function/eb30af1d9d654ebc98287ca25f231bf6/) "
            "within a Function instance which is applied to the stored document when "
            "considering storage time."
        ),
    )
    draft = models.BooleanField(
        default=False,
        verbose_name=_("draft"),
        help_text=_(
            "Is this document a draft or not. Drafts can be modified by a user."
        ),
    )
    locked_after = models.DateTimeField(
        blank=True,
        null=True,
        verbose_name=_("locked after"),
        help_text=_(
            "Date and time after which this document cannot be modified, except for "
            "deleting. This field should be filled by the calling service if it knows "
            "e.g. that a certain application has a deadline."
        ),
    )
    metadata = models.JSONField(
        default=dict,
        blank=True,
        # null=True,  # TODO should we allow null?
        encoder=DjangoJSONEncoder,
        verbose_name=_("metadata"),
        help_text=_(
            "Key-value pairs given by the calling service. These fields should enable "
            "the service to store some relevant information which it can use to "
            "filter/sort documents, e.g. the handler of the document."
        ),
    )
    content = fields.EncryptedTextField(
        help_text=_("Encrypted content of the document.")
    )
    status = models.CharField(
        max_length=255,
        blank=True,
        help_text=_(
            "Status information given by the owning service. "
            "Could be e.g. some constant string."
        ),
    )
    type = models.CharField(
        max_length=255,
        blank=True,
        verbose_name=_("type"),
        help_text=_(
            "Type information given by the owning service. "
            "Could be e.g. the type of the document."
        ),
    )

    class Meta:
        verbose_name = _("document")
        verbose_name_plural = _("documents")
        default_related_name = "documents"
        indexes = [
            models.Index(fields=["created_at"], name="document_created_at_idx"),
            models.Index(fields=["updated_at"], name="document_updated_at_idx"),
            models.Index(fields=["business_id"], name="document_business_id_idx"),
            models.Index(fields=["transaction_id"], name="document_transaction_id_idx"),
            models.Index(fields=["draft"], name="document_draft_idx"),
            models.Index(fields=["locked_after"], name="document_locked_after_idx"),
            GinIndex(fields=["metadata"], name="document_metadata_idx"),
            models.Index(fields=["status"], name="document_status_idx"),
            models.Index(fields=["type"], name="document_type_idx"),
        ]

    def __str__(self):
        return f"Document {self.pk}"
