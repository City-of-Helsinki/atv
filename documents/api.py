from rest_framework import status, viewsets
from rest_framework.response import Response


class DocumentViewSet(viewsets.ViewSet):
    def create(self, request):
        return Response({"status": "CREATED"}, status=status.HTTP_201_CREATED)
