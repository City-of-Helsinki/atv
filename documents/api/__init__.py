from .viewsets import AttachmentViewSet, DocumentViewSet

__all__ = [
    "AttachmentViewSet",
    "DocumentViewSet",
]
