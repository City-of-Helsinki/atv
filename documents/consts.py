VALID_OWNER_PATCH_FIELDS = {"draft", "metadata", "content", "attachments"}
