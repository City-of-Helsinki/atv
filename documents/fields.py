import json
from io import BytesIO

from Crypto.Cipher import AES
from django.db import models
from encrypted_fields.fields import EncryptedFieldMixin

from atv.settings import FIELD_ENCRYPTION_KEYS


class EncryptedJSONField(EncryptedFieldMixin, models.JSONField):
    def decrypt(self, value):
        text = super().decrypt(value)
        return json.loads(text)

    def get_db_prep_save(self, value, connection):
        if self.empty_strings_allowed and value == bytes():
            value = ""
        # Instead of using db_prep_value use json.dumps() to get a json formatted object
        value = json.dumps(value)
        if value is not None:
            encrypted_value = self.encrypt(value)
            return connection.Database.Binary(encrypted_value)


class EncryptedFileField(models.FileField):
    def pre_save(self, model_instance, add):
        file = getattr(model_instance, self.attname)
        if file and not file._committed:
            # Commit the file to storage prior to saving the model
            file.save(file.name, self.encrypt_file(file.file), save=False)
        return file

    @staticmethod
    def encrypt_file(file):
        data_to_encrypt = file.read()
        cipher = AES.new(bytes.fromhex(FIELD_ENCRYPTION_KEYS[0]), AES.MODE_GCM)
        nonce = cipher.nonce
        cypher_text, tag = cipher.encrypt_and_digest(data_to_encrypt)
        return BytesIO(nonce + tag + cypher_text)
