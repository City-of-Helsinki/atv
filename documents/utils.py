from io import BytesIO

from Cryptodome.Cipher import AES
from django.conf import settings
from django.core.files import File
from pyclamd import pyclamd

from atv.exceptions import MaliciousFileException
from atv.settings import FIELD_ENCRYPTION_KEYS


def get_attachment_file_path(instance, filename):
    """File will be uploaded to MEDIA_ROOT/ATTACHMENT_MEDIA_DIR/<document_id>/<filename>"""
    return f"{settings.ATTACHMENT_MEDIA_DIR}/{instance.document.id}/{filename}"


def get_document_attachment_directory_path(instance):
    """Get the root directory for a document's attachments.

    :type instance: documents.models.Document
    """
    return f"{settings.ATTACHMENT_MEDIA_DIR}/{instance.id}/"


def get_decrypted_file(file, file_name):
    nonce = file[:16]
    # Perform same nonce checks here as Pycryptodome, so we can raise a more user
    # friendly error message
    if not isinstance(nonce, (bytes, bytearray, memoryview)) or len(nonce) != 16:
        raise ValueError("Data is corrupted.")
    tag = file[16:32]
    cypher_text = file[32:]
    counter = 0
    num_keys = len(FIELD_ENCRYPTION_KEYS)
    while counter < num_keys:
        cipher = AES.new(
            bytes.fromhex(FIELD_ENCRYPTION_KEYS[counter]), AES.MODE_GCM, nonce=nonce
        )
        try:
            plaintext = cipher.decrypt_and_verify(cypher_text, tag)
        except ValueError:
            counter += 1
            continue
        return File(BytesIO(plaintext), name=file_name)
    raise ValueError("AES Key incorrect or data is corrupted")


# TODO: Consider scanning files on download as well to improve chance of catching most recent threats to protect users
#   if clamav virus databases didn't include the virus' profile at the time of upload
# Note this function is mocked in testing because there is no clamav connection during pipeline testing
def virus_scan_attachment_file(file_data):
    cd = pyclamd.ClamdNetworkSocket(host=settings.CLAMAV_HOST)
    if cd.scan_stream(file_data) is not None:
        raise MaliciousFileException
