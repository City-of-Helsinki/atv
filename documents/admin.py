from django.contrib import admin
from django.urls import reverse
from django.utils.html import format_html
from django.utils.translation import gettext_lazy as _

from documents.models import Attachment, Document


@admin.register(Document)
class DocumentAdmin(admin.ModelAdmin):
    list_display = ("pk", "get_service_name", "draft", "created_at", "updated_at")
    search_fields = ("service__name", "type", "status", "business_id")
    list_filter = ("draft",)
    autocomplete_fields = ("service", "user")
    list_select_related = ("service",)

    @admin.display(description=_("service"), ordering="-service__name")
    def get_service_name(self, obj):
        return obj.service.name


@admin.register(Attachment)
class AttachmentAdmin(admin.ModelAdmin):
    list_display = (
        "pk",
        "filename",
        "media_type",
        "get_size_in_mb",
        "get_document",
        "created_at",
        "updated_at",
    )
    search_fields = (
        "filename",
        "document__id",
    )
    list_filter = ("media_type",)
    autocomplete_fields = ("document",)
    readonly_fields = (
        "filename",
        "media_type",
        "size",
    )

    @admin.display(description=_("size (MB)"))
    def get_size_in_mb(self, obj):
        if not obj.size:
            return "-"
        size_mb = round(float(obj.size) / (1024 ** 2), 2)
        return f"{size_mb}MB"

    @admin.display(description=_("Document"))
    def get_document(self, obj):
        link = reverse("admin:documents_document_change", args=[obj.document.id])
        return format_html('<a href="{}">{}</a>', link, str(obj.document))
