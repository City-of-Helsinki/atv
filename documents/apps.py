from django.apps import AppConfig
from django.utils.translation import gettext_lazy as _


class DocumentsConfig(AppConfig):
    default_auto_field = "django.db.models.BigAutoField"
    name = "documents"
    verbose_name = _("documents")

    def ready(self):
        import documents.signals  # noqa
