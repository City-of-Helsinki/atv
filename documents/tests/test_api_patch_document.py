import json
from unittest import mock
from uuid import uuid4

import pytest
from dateutil.relativedelta import relativedelta
from dateutil.utils import today
from django.core.files.uploadedfile import SimpleUploadedFile
from freezegun import freeze_time
from guardian.shortcuts import assign_perm
from rest_framework import status
from rest_framework.reverse import reverse

from atv.tests.factories import GroupFactory
from audit_log.models import AuditLogEntry
from documents.models import Document, StatusHistory
from documents.tests.factories import DocumentFactory
from documents.tests.test_api_create_document import VALID_DOCUMENT_DATA
from services.enums import ServicePermissions
from services.tests.utils import get_user_service_client
from users.models import User
from utils.exceptions import get_error_response

VALID_OWNER_DOCUMENT_DATA = {
    "draft": False,
    "metadata": json.dumps({"created_by": "alex", "testing": True}),
    "content": json.dumps(
        {
            "formData": {
                "firstName": "Dolph",
                "lastName": "Lundgren",
                "birthDate": "3.11.1957",
            },
            "reasonForApplication": "No reason, just testing",
        }
    ),
}


# OWNER-RELATED ACTIONS


@freeze_time("2021-06-30T12:00:00+03:00")
def test_update_document_owner(user, snapshot):
    document = DocumentFactory(
        id="2d2b7a36-a306-4e35-990f-13aea04263ff",
        draft=True,
        user=user,
        status="handled",
        type="mysterious form",
        transaction_id="cf0a341b-6bfd-4f59-8d7c-87bf62ba837b",
        business_id="1234567-8",
        tos_function_id="f917d43aab76420bb2ec53f6684da7f7",
        tos_record_id="89837a682b5d410e861f8f3688154163",
    )
    assert document.attachments.count() == 0

    api_client = get_user_service_client(user, document.service)

    data = {
        **VALID_OWNER_DOCUMENT_DATA,
        "attachments": [
            SimpleUploadedFile(
                "document1.pdf", b"file_content", content_type="application/pdf"
            ),
        ],
    }
    with mock.patch(
        "documents.serializers.attachment.virus_scan_attachment_file", return_value=None
    ):
        response = api_client.patch(
            reverse("documents-detail", args=[document.id]), data, format="multipart"
        )

    assert Document.objects.count() == 1
    assert document.attachments.count() == 1

    body = response.json()
    attachment = document.attachments.first()

    assert body.pop("attachments", []) == [
        {
            "created_at": "2021-06-30T12:00:00+03:00",
            "filename": "document1.pdf",
            "href": f"http://testserver/v1/documents/2d2b7a36-a306-4e35-990f-13aea04263ff/attachments/{attachment.id}/",
            "id": attachment.id,
            "media_type": "application/pdf",
            "size": 12,
            "updated_at": "2021-06-30T12:00:00+03:00",
        }
    ]

    assert response.status_code == status.HTTP_200_OK
    assert body.pop("user_id", None) == str(document.user.uuid) == str(user.uuid)

    snapshot.assert_match(body)


@freeze_time("2021-06-30T12:00:00+03:00")
def test_update_document_owner_someone_elses_document(
    user,
):
    document = DocumentFactory(id="2d2b7a36-a306-4e35-990f-13aea04263ff", draft=True)
    api_client = get_user_service_client(user, document.service)
    # Check that the owner of the document is different than the one
    # making the request
    assert document.user != user

    response = api_client.patch(
        reverse("documents-detail", args=[document.id]),
        {},  # No need to have any actual data
    )

    body = response.json()

    assert response.status_code == status.HTTP_404_NOT_FOUND
    assert body == get_error_response(
        "NOT_FOUND", "No Document matches the given query."
    )


@freeze_time("2021-06-30T12:00:00+03:00")
def test_update_document_owner_non_draft(user):
    document = DocumentFactory(
        id="2d2b7a36-a306-4e35-990f-13aea04263ff", draft=False, user=user
    )
    api_client = get_user_service_client(user, document.service)

    response = api_client.patch(
        reverse("documents-detail", args=[document.id]),
        {},  # No need to have any actual data
    )

    body = response.json()

    assert response.status_code == status.HTTP_403_FORBIDDEN
    assert body == get_error_response(
        "DOCUMENT_LOCKED", "Unable to modify document - it's no longer a draft"
    )


@freeze_time("2021-06-30T12:00:00")
def test_update_document_owner_after_lock_date(user):
    document = DocumentFactory(
        id="2d2b7a36-a306-4e35-990f-13aea04263ff",
        locked_after=today() - relativedelta(days=1),
        draft=True,
        user=user,
    )
    api_client = get_user_service_client(user, document.service)

    response = api_client.patch(
        reverse("documents-detail", args=[document.id]),
        {},  # No need to have any actual data
    )

    body = response.json()

    assert response.status_code == status.HTTP_403_FORBIDDEN
    assert body == get_error_response(
        "DOCUMENT_LOCKED", "Unable to modify document - it's no longer a draft"
    )


@freeze_time("2021-06-30T12:00:00")
def test_update_document_owner_invalid_fields(user):
    document = DocumentFactory(
        id="2d2b7a36-a306-4e35-990f-13aea04263ff", draft=True, user=user
    )
    api_client = get_user_service_client(user, document.service)

    response = api_client.patch(
        reverse("documents-detail", args=[document.id]),
        {
            "tos_function_id": "f917d43aab76420bb2ec53f6684da7f7",
            "tos_record_id": "89837a682b5d410e861f8f3688154163",
        },  # No need to have any actual data
    )

    body = response.json()

    assert response.status_code == status.HTTP_400_BAD_REQUEST

    assert len(body.get("errors")) == 1
    assert body.get("errors", [])[0].get("code") == "INVALID_FIELD"
    assert "Got invalid input fields" in body.get("errors", [])[0].get("message")
    assert "tos_function_id" in body.get("errors", [])[0].get("message")
    assert "tos_record_id" in body.get("errors", [])[0].get("message")


# STAFF-RELATED ACTION


@freeze_time("2021-06-30T12:00:00+03:00")
def test_update_document_staff(user, service, snapshot):
    api_client = get_user_service_client(user, service)
    group = GroupFactory()
    assign_perm(ServicePermissions.VIEW_DOCUMENTS.value, group, service)
    assign_perm(ServicePermissions.CHANGE_DOCUMENTS.value, group, service)
    assign_perm(ServicePermissions.ADD_ATTACHMENTS.value, group, service)
    user.groups.add(group)

    document = DocumentFactory(
        id="2d2b7a36-a306-4e35-990f-13aea04263ff",
        draft=True,
        service=service,
        status="handled",
    )
    assert Document.objects.count() == 1

    data = {
        **VALID_DOCUMENT_DATA,
        "attachments": [
            SimpleUploadedFile(
                "document1.pdf", b"file_content", content_type="application/pdf"
            ),
        ],
    }
    with mock.patch(
        "documents.serializers.attachment.virus_scan_attachment_file", return_value=None
    ):
        response = api_client.patch(
            reverse("documents-detail", args=[document.id]), data, format="multipart"
        )

    assert Document.objects.count() == 1
    assert document.attachments.count() == 1

    body = response.json()
    attachment = document.attachments.first()

    assert body.pop("attachments", []) == [
        {
            "created_at": "2021-06-30T12:00:00+03:00",
            "filename": "document1.pdf",
            "href": f"http://testserver/v1/documents/2d2b7a36-a306-4e35-990f-13aea04263ff/attachments/{attachment.id}/",
            "id": attachment.id,
            "media_type": "application/pdf",
            "size": 12,
            "updated_at": "2021-06-30T12:00:00+03:00",
        }
    ]

    assert response.status_code == status.HTTP_200_OK
    assert body.pop("user_id", None) == str(document.user.uuid)
    snapshot.assert_match(body)


@freeze_time("2021-06-30T12:00:00+03:00")
def test_update_document_staff_non_draft(user, service, snapshot):
    api_client = get_user_service_client(user, service)
    group = GroupFactory()
    assign_perm(ServicePermissions.VIEW_DOCUMENTS.value, group, service)
    assign_perm(ServicePermissions.CHANGE_DOCUMENTS.value, group, service)
    user.groups.add(group)

    document = DocumentFactory(
        id="2d2b7a36-a306-4e35-990f-13aea04263ff",
        service=service,
        draft=False,
        status="handled",
    )

    data = {**VALID_DOCUMENT_DATA}

    response = api_client.patch(reverse("documents-detail", args=[document.id]), data)

    body = response.json()

    assert response.status_code == status.HTTP_200_OK
    assert body.pop("user_id", None) == str(document.user.uuid)
    snapshot.assert_match(body)


@freeze_time("2021-06-30T12:00:00")
def test_update_document_staff_after_lock_date(user, service):
    api_client = get_user_service_client(user, service)
    group = GroupFactory()
    assign_perm(ServicePermissions.VIEW_DOCUMENTS.value, group, service)
    assign_perm(ServicePermissions.CHANGE_DOCUMENTS.value, group, service)
    user.groups.add(group)

    document = DocumentFactory(
        id="2d2b7a36-a306-4e35-990f-13aea04263ff",
        locked_after=today() - relativedelta(days=1),
        service=service,
    )

    response = api_client.patch(
        reverse("documents-detail", args=[document.id]),
        {},  # No need to have any actual data
    )

    body = response.json()

    assert response.status_code == status.HTTP_403_FORBIDDEN
    assert body == get_error_response(
        "DOCUMENT_LOCKED", "Unable to modify document - it's no longer a draft"
    )


@freeze_time("2021-06-30T12:00:00")
def test_update_document_staff_another_service(user, service):
    api_client = get_user_service_client(user, service)
    group = GroupFactory()
    assign_perm(ServicePermissions.VIEW_DOCUMENTS.value, group, service)
    assign_perm(ServicePermissions.CHANGE_DOCUMENTS.value, group, service)
    user.groups.add(group)

    document = DocumentFactory(id="2d2b7a36-a306-4e35-990f-13aea04263ff")

    response = api_client.patch(
        reverse("documents-detail", args=[document.id]),
        {},  # No need to have any actual data
    )

    body = response.json()

    assert response.status_code == status.HTTP_404_NOT_FOUND
    assert body == get_error_response(
        "NOT_FOUND", "No Document matches the given query."
    )


@freeze_time("2021-06-30T12:00:00")
def test_add_user_anonymous_document(service_api_client):
    data = {
        **VALID_DOCUMENT_DATA,
    }

    response = service_api_client.post(
        reverse("documents-list"), data, format="multipart"
    )
    assert response.status_code == status.HTTP_201_CREATED
    document_id = response.json().get("id")

    user_id = "6345c12c-36c8-4e81-bd18-d66e9b1f754d"
    response = service_api_client.patch(
        reverse("documents-detail", args=[document_id]), {"user_id": user_id}
    )
    assert response.status_code == status.HTTP_200_OK
    assert User.objects.filter(uuid=user_id).exists() is True

    body = response.json()
    assert body.get("user_id") == user_id


@freeze_time("2021-06-30T12:00:00")
def test_update_document_user(service_api_client):
    user_id = "6345c12c-36c8-4e81-bd18-d66e9b1f754d"
    new_user_id = "6345c12c-36c8-4e81-bd18-d66e9b1f754b"
    data = {**VALID_DOCUMENT_DATA, "user_id": user_id}

    response = service_api_client.post(
        reverse("documents-list"), data, format="multipart"
    )
    body = response.json()

    assert response.status_code == status.HTTP_201_CREATED
    assert User.objects.filter(uuid=user_id).exists() is True
    assert body.get("user_id") == user_id

    document_id = body.get("id")
    response = service_api_client.patch(
        reverse("documents-detail", args=[document_id]), {"user_id": new_user_id}
    )
    assert response.status_code == status.HTTP_200_OK
    assert str(Document.objects.get(id=document_id).user.uuid) == new_user_id


@freeze_time("2021-06-30T12:00:00")
def test_update_document_owner_user_id(user):
    document = DocumentFactory(
        id="2d2b7a36-a306-4e35-990f-13aea04263ff",
        draft=True,
        user=user,
        status="handled",
        type="mysterious form",
        transaction_id="cf0a341b-6bfd-4f59-8d7c-87bf62ba837b",
        business_id="1234567-8",
        tos_function_id="f917d43aab76420bb2ec53f6684da7f7",
        tos_record_id="89837a682b5d410e861f8f3688154163",
    )

    assert Document.objects.count() == 1
    api_client = get_user_service_client(user, document.service)

    user_id = "6345c12c-36c8-4e81-bd18-d66e9b1f754d"
    data = {**VALID_OWNER_DOCUMENT_DATA, "user_id": user_id}
    response = api_client.patch(
        reverse("documents-detail", args=[document.id]), data, format="multipart"
    )
    assert response.status_code == status.HTTP_400_BAD_REQUEST
    assert not User.objects.filter(uuid=user_id).exists()

    assert response.json() == get_error_response(
        code="INVALID_FIELD", detail="Got invalid input fields: user_id."
    )


@freeze_time("2021-06-30T12:00:00")
def test_update_deletable_field(user, service):
    api_client = get_user_service_client(user, service)
    group = GroupFactory()
    assign_perm(ServicePermissions.VIEW_DOCUMENTS.value, group, service)
    assign_perm(ServicePermissions.CHANGE_DOCUMENTS.value, group, service)
    user.groups.add(group)

    document = DocumentFactory(
        id="2d2b7a36-a306-4e35-990f-13aea04263ff", service=service, deletable=True
    )
    assert Document.objects.count() == 1

    # Allow changing deletable field from True to False
    response = api_client.patch(
        reverse("documents-detail", args=[document.id]),
        {"deletable": False},
    )
    assert response.status_code == status.HTTP_200_OK
    assert response.json()["deletable"] is False

    # Changing deletable to True from False isn't allowed
    response = api_client.patch(
        reverse("documents-detail", args=[document.id]),
        {"deletable": True},
    )
    assert response.status_code == status.HTTP_403_FORBIDDEN


@freeze_time("2021-06-30T12:00:00")
def test_create_status_history(service_api_client):
    data = {
        **VALID_DOCUMENT_DATA,
    }

    response = service_api_client.post(
        reverse("documents-list"), data, format="multipart"
    )
    assert response.status_code == status.HTTP_201_CREATED
    document_id = response.json().get("id")

    # Patch shouldn't create status history object
    response = service_api_client.patch(
        reverse("documents-detail", args=[document_id]), {"type": "new type"}
    )
    assert response.status_code == status.HTTP_200_OK
    assert StatusHistory.objects.count() == 0

    # Creates status history object when status field changes
    response = service_api_client.patch(
        reverse("documents-detail", args=[document_id]),
        {"status": "created status history object"},
    )

    assert response.status_code == status.HTTP_200_OK
    assert StatusHistory.objects.count() == 1


# OTHER STUFF


@freeze_time("2021-06-30T12:00:00")
def test_update_document_not_found(superuser_api_client):
    response = superuser_api_client.patch(
        reverse("documents-detail", args=[uuid4()]),
        {},  # No need to have any actual data
    )

    body = response.json()

    assert response.status_code == status.HTTP_404_NOT_FOUND
    assert body == get_error_response(
        "NOT_FOUND", "No Document matches the given query."
    )


@pytest.mark.parametrize("attachments", [0, 1, 2])
@pytest.mark.parametrize(
    "ip_address", ["213.255.180.34", "2345:0425:2CA1::0567:5673:23b5"]
)
def test_audit_log_is_created_when_patching(user, attachments, ip_address):
    document = DocumentFactory(draft=True, user=user)
    api_client = get_user_service_client(user, document.service)
    data = {**VALID_OWNER_DOCUMENT_DATA}

    if attachments:
        data["attachments"] = [
            SimpleUploadedFile(
                f"document{i}.pdf", b"file_content", content_type="application/pdf"
            )
            for i in range(attachments)
        ]
    with mock.patch(
        "documents.serializers.attachment.virus_scan_attachment_file", return_value=None
    ):
        api_client.patch(
            reverse("documents-detail", args=[document.id]),
            data,
            HTTP_X_FORWARDED_FOR=ip_address,
        )

    assert (
        AuditLogEntry.objects.filter(
            message__audit_event__target__type="Document",
            message__audit_event__target__id=str(document.pk),
            message__audit_event__target__lookup_field="pk",
            message__audit_event__target__endpoint="Document Instance",
            message__audit_event__operation="UPDATE",
            message__audit_event__actor__ip_address=ip_address,
        ).count()
        == 1
    )
