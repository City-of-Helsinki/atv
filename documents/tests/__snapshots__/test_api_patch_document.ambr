# serializer version: 1
# name: test_update_document_owner
  dict({
    'business_id': '1234567-8',
    'content': dict({
      'formData': dict({
        'birthDate': '3.11.1957',
        'firstName': 'Dolph',
        'lastName': 'Lundgren',
      }),
      'reasonForApplication': 'No reason, just testing',
    }),
    'content_schema_url': None,
    'created_at': '2021-06-30T12:00:00+03:00',
    'deletable': True,
    'delete_after': None,
    'document_language': None,
    'draft': False,
    'human_readable_type': dict({
    }),
    'id': '2d2b7a36-a306-4e35-990f-13aea04263ff',
    'locked_after': None,
    'metadata': dict({
      'created_by': 'alex',
      'testing': True,
    }),
    'service': 'service 160',
    'status': dict({
      'status_display_values': dict({
      }),
      'timestamp': '2021-06-30T12:00:00+03:00',
      'value': 'handled',
    }),
    'status_histories': list([
    ]),
    'tos_function_id': 'f917d43aab76420bb2ec53f6684da7f7',
    'tos_record_id': '89837a682b5d410e861f8f3688154163',
    'transaction_id': 'cf0a341b-6bfd-4f59-8d7c-87bf62ba837b',
    'type': 'mysterious form',
    'updated_at': '2021-06-30T12:00:00+03:00',
  })
# ---
# name: test_update_document_staff
  dict({
    'business_id': '1234567-8',
    'content': dict({
      'formData': dict({
        'birthDate': '3.11.1957',
        'firstName': 'Dolph',
        'lastName': 'Lundgren',
      }),
      'reasonForApplication': 'No reason, just testing',
    }),
    'content_schema_url': 'https://schema.fi',
    'created_at': '2021-06-30T12:00:00+03:00',
    'deletable': False,
    'delete_after': None,
    'document_language': 'en',
    'draft': True,
    'human_readable_type': dict({
      'en': 'Mysterious Form',
    }),
    'id': '2d2b7a36-a306-4e35-990f-13aea04263ff',
    'locked_after': None,
    'metadata': dict({
      'created_by': 'alex',
      'testing': True,
    }),
    'service': 'service 166',
    'status': dict({
      'status_display_values': dict({
        'fi': 'Käsitelty',
      }),
      'timestamp': '2021-06-30T12:00:00+03:00',
      'value': 'handled',
    }),
    'status_histories': list([
    ]),
    'tos_function_id': 'f917d43aab76420bb2ec53f6684da7f7',
    'tos_record_id': '89837a682b5d410e861f8f3688154163',
    'transaction_id': 'cf0a341b-6bfd-4f59-8d7c-87bf62ba837b',
    'type': 'mysterious form',
    'updated_at': '2021-06-30T12:00:00+03:00',
  })
# ---
# name: test_update_document_staff_non_draft
  dict({
    'attachments': list([
    ]),
    'business_id': '1234567-8',
    'content': dict({
      'formData': dict({
        'birthDate': '3.11.1957',
        'firstName': 'Dolph',
        'lastName': 'Lundgren',
      }),
      'reasonForApplication': 'No reason, just testing',
    }),
    'content_schema_url': 'https://schema.fi',
    'created_at': '2021-06-30T12:00:00+03:00',
    'deletable': False,
    'delete_after': None,
    'document_language': 'en',
    'draft': False,
    'human_readable_type': dict({
      'en': 'Mysterious Form',
    }),
    'id': '2d2b7a36-a306-4e35-990f-13aea04263ff',
    'locked_after': None,
    'metadata': dict({
      'created_by': 'alex',
      'testing': True,
    }),
    'service': 'service 168',
    'status': dict({
      'status_display_values': dict({
        'fi': 'Käsitelty',
      }),
      'timestamp': '2021-06-30T12:00:00+03:00',
      'value': 'handled',
    }),
    'status_histories': list([
    ]),
    'tos_function_id': 'f917d43aab76420bb2ec53f6684da7f7',
    'tos_record_id': '89837a682b5d410e861f8f3688154163',
    'transaction_id': 'cf0a341b-6bfd-4f59-8d7c-87bf62ba837b',
    'type': 'mysterious form',
    'updated_at': '2021-06-30T12:00:00+03:00',
  })
# ---
